python/cudf/cudf/dataframe/columnops.py

```diff
@@ -113,6 +114,53 @@ def apply_boolean_mask(self, mask):
     def fillna(self, fill_value, inplace):
         raise NotImplementedError
 
+    def astype(self, dtype, **kwargs):
+        if pd.api.types.is_string_dtype(dtype):
+            if pd.api.types.is_categorical_dtype(dtype):
+                return self.as_categorical_column(dtype, **kwargs)
+            return self.as_string_column(dtype, **kwargs)
+
+        elif np.issubdtype(dtype, np.datetime64):
+            return self.as_datetime_column(dtype, **kwargs)
+
+        else:
+            return self.as_numerical_column(dtype, **kwargs)
+
+    def as_categorical_column(self, dtype, **kwargs):
+        if "ordered" in kwargs:
+            ordered = kwargs["ordered"]
+        else:
+            ordered = False
+
+        sr = cudf.Series(self)
+        labels, cats = sr.factorize()
+
+        # string columns include null index in factorization; remove:
+        if (
+            pd.api.types.is_string_dtype(self.dtype)
+            and not pd.api.types.is_categorical_dtype(self.dtype)
```

Comment:
Both of these functions are actually really expensive because they're designed to check for array-like objects. Could we use `pandas.api.types.pandas_dtype(self.dtype).type is yada_yada` instead?

Reply:
What order of magnitude should be considered "expensive"? I didn't find them to be terribly different in performance:

```
In [13]: %timeit pd.api.types.pandas_dtype("category").type is pd.CategoricalDtype                                                                                                            
1.33 µs ± 1.17 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: %timeit pd.api.types.is_categorical_dtype("category")                                                                                                                                 
1.91 µs ± 3.44 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
This was what we were seeing I think:
```
In [3]: test = pd.CategoricalDtype()

In [4]: %timeit pd.api.types.pandas_dtype(test).type is pd.CategoricalDtype
511 ns ± 1.18 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [5]: %timeit pd.api.types.is_categorical_dtype(test)
2.55 µs ± 8.59 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
OK but:

```
In [7]: test = pd.CategoricalDtype()                                                                                                                                                                            

In [8]: pd.api.types.pandas_dtype(test).type is pd.CategoricalDtype                                                                                                                                             
Out[8]: False

In [9]: pd.api.types.is_categorical_dtype(test)                                                                                                                                                                 
Out[9]: True
```

😁

Reply:
I'd ask that we keep the usage of `is_categorical_dtype`. It's more general and does exactly what we want here:

> Check whether an array-like **or dtype** is of the Categorical dtype.

Was it found to be a bottleneck specifically somewhere?

Reply:
It should have been `is pd.core.dtypes.dtypes.CategoricalDtypeType` but yes it was found to be a bottleneck in `get_dummies` related work where it was taking significant time when handling things like shallow copies and even deep copies of columns.